Definite assignment, initial nullable, IOperation, and some cleanup

333fred opened the pull request:
Implements support for definite assignment, initial nullable support, initial IOperation support, and does a bit of prototype comment cleanup. Next week I'll add some additional commits cleaning up the rest of the prototype comments in this branch, they're mainly around adding tests for dynamic to be sure of behavior, ref escape behavior custom interpolated string handler arguments, and condensing message numbers for merge.

Test plan: https://github.com/dotnet/roslyn/issues/51499

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -1149,23 +1149,47 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
+                case { Construction: var construction }:
+                    VisitRvalue(construction);
+                    var beforePartsState = State.Clone();
+                    visitParts();
```
Feels like we should join the state before and after each part in this case when `UsesBoolReturns` is true. I think this would affect nullability analysis and maybe also region analysis.
Also, if I understood coorectly, the handler constructor having a success value is independant of the append calls having a success value. We probably need to handle those permutations.

333fred replied:
I don't think this can actually have an effect.

jcouv replied:
I think it should be possible to observe. Try putting assignments in the holes (`$"{x = null} {x = string.Empty}"`) and observe `x` afterwards.

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs:
```diff
@@ -243,6 +243,17 @@ private void VisitBinaryOperatorChildren(BoundBinaryOperatorBase node)
             }
+
+            public override BoundNode? VisitInterpolatedString(BoundInterpolatedString node)
+            {
```
nit: extra blank line below

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -1149,23 +1149,47 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
         }
 
-        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node)
+        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node, InterpolatedStringHandlerData? nullableData)
```
Do we need to account for some expressions being evaluated early due to argument attribute?

333fred replied:
Huh? That's not a thing.

333fred replied:
To be specific, they're not evaluated early. They're evaluated at exactly the same time as they were before.

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -1149,23 +1149,47 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
         }
 
-        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node)
+        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node, InterpolatedStringHandlerData? nullableData)
```
>nullableData [](http://example.com/codeflow?start=121&length=12)

nit: consider `data` like in other places

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -1149,23 +1149,47 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
+                    visitParts();
+                    Join(ref State, ref beforePartsState);
+                    break;
             }
```
nit: consider adding a default case with a `throw`

333fred replied:
Not possible to hit. The 3 cases given are fully exhaustive.

jcouv commented:
>         public void DefiniteAssignment(bool useBoolReturns, bool trailingOutParameter)

It would be good to include a region analysis test as well.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10243 in 1db9ee9. [](commit_id = 1db9ee9e624d3a853cfad09991c82c2b175e2b68, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -2110,7 +2110,6 @@ private static ParameterSymbol GetParameter(int argIndex, MemberAnalysisResult r
             }
 
             // Otherwise, prefer methods with 'val' parameters over 'in' parameters and over 'ref' parameters when the argument is an interpolated string handler.
-            // PROTOTYPE(interp-string): Document in the spec
```
Is there a corresponding PR for spec changes?

333fred replied:
Merged some time ago.

jcouv commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:
```diff
@@ -1416,7 +1443,314 @@ public void M(bool a, string b, string c, string c2, string p)
+        }
+
+        [Fact]
+        public void InterpolatedStringHandlerConversion_07()
```
It would be good to include a test of conversion in call with argument attribute, as smoke test. I understand the full implementation will come later.

333fred replied:
We have lots of smoke tests with argument attributes via the test hook. I don't think this is necessary.

jcouv replied:
I understand the test hook verifies we don't crash and the tree is somewhat well-formed, but we aren't capturing the IOp tree anywhere. Seems easy to add for peace of mind given this is a central scenario.

333fred replied:
It's not going to be any different. We're not doing anything with the construction parameter of interpolation data today, which is the only place where there is going to be any difference. I think IOp hook is good enough for now.

jcouv commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1973,29 +1988,140 @@ internal IOperation CreateBoundTupleOperation(BoundTupleExpression boundTupleExp
+                            (value, alignment, format) = getCallInfo(call.Arguments, call.ArgumentNamesOpt, currentPosition);
+                            break;
+
+                        case BoundDynamicInvocation dynamicInvocation:
```
I couldn't identify a test that hits this case.

333fred replied:
We do not. It's on my plan did Tuesday.

333fred replied:
And I did not remove the prototype comment related to dynamic because of this.

jcouv replied:
> It's on my plan for Tuesday.

Sounds good. Thanks
My understanding is that we don't support dynamic scenario in binding or lowering, so I was surprised to see some support in IOp.

333fred replied:
We don't support emitting code, but that doesn't mean the IOp tree creation should fail.

jcouv reviewed (commented):
Done with review pass (iteration 5)

333fred commented:
>         public void DefiniteAssignment(bool useBoolReturns, bool trailingOutParameter)

Is this something you want to block on, or can it be added later?

---
In reply to: [874428443](https://github.com/dotnet/roslyn/pull/54585#issuecomment-874428443) [](http://example.com/codeflow?ancestors=874428443,874428443)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10243 in 1db9ee9. [](commit_id = 1db9ee9e624d3a853cfad09991c82c2b175e2b68, deletion_comment = False)

333fred commented:
@jcouv addressed feedback. I'd suggest reviewing commits 6 and 7 one at a time.

---
In reply to: [699512610](https://github.com/dotnet/roslyn/pull/54585#issuecomment-699512610) [](http://example.com/codeflow?ancestors=699512610,699512610)

jcouv commented:
@333fred There seems to be legitimate test failures

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -1149,23 +1149,65 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
+                    break;
+                case { UsesBoolReturns: var usesBoolReturns, Construction: var construction }:
+                    VisitRvalue(construction);
+                    var beforePartsState = State.Clone();
```
Feels like we should have different behavior depending on value of `HasTrailingHandlerValidityParameter`, but that value is ignored here.
If it is `false`, we're guaranteed to execute the first append call. If it's `true`, we may not execute any append.
The behavior as implemented seems to assume that `HasTrailingHandlerValidityParameter` was `true`.

Once we have that, we should take another look at the second case and see whether it's still necessary (I suspect it should just fall out from the last case using `false` and `false`).

333fred replied:
Given that the behavior in this section is predicated on there being some amount of conditional evaluation, I think it's cleaner to keep it separate.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:
```diff
@@ -393,6 +393,9 @@ public override BoundNode VisitObjectOrCollectionValuePlaceholder(BoundObjectOrC
         public override BoundNode VisitInterpolatedStringArgumentPlaceholder(BoundInterpolatedStringArgumentPlaceholder node)
             => PlaceholderReplacement(node);
 
+        public override BoundNode? VisitInterpolatedStringHandlerPlaceholder(BoundInterpolatedStringHandlerPlaceholder node)
```
nit: we should similarly add a check in `LocalRewritingValidator` below for `BoundInterpolatedStringHandlerPlaceholder`.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:
```diff
@@ -3347,6 +3403,38 @@ private static bool CheckValEscape(ImmutableArray<BoundExpression> expressions,
+                // The interpolation component is always the first argument to the method, and it was not passed by name
+                // so there can be no reordering.
+                var argument = call.Arguments[0];
+                var success = isRef
```
For expressions passed via the holes, we check `GetRefEscape` or `GetValEscape` depending on , but for arguments passed via argument attribute we only check the val-escape (`ValSafeToEscape`). Just want to confirm that's intentional/correct.

jcouv replied:
Mm... it looks like this method is only ever called with `isRef` `false`.

333fred replied:
I'm also missing a check on the constructor, I'll add that as well.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -1020,7 +1020,7 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
             var receiver = ReplaceTypeOrValueReceiver(methodGroup.Receiver, !method.RequiresInstanceReceiver && !invokedAsExtensionMethod, diagnostics);
 
-            this.CoerceArguments(methodResult, analyzedArguments.Arguments, diagnostics, receiver?.Type, receiver?.GetRefKind());
+            this.CoerceArguments(methodResult, analyzedArguments.Arguments, diagnostics, receiver?.Type, receiver?.GetRefKind(), method.RequiresInstanceReceiver ? GetValEscape(receiver, LocalScopeDepth) : Binder.ExternalScope);
```
If `receiver?.GetRefKind()` is `ref`, should we use `GetRefEscape` instead of `GetValEscape`?

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:
```diff
@@ -1149,23 +1150,84 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
+                    }
+                    else
+                    {
+                        Join(ref State, ref beforePartsState);
```
I didn't understand this case. Why do we need to `Join` when none of the evaluations are conditional?

333fred replied:
It is not. `hasTrailingValidityParameter` will be true in that case.

jcouv replied:
Thanks for the clarification

chsienki replied:
Nit: this also confused me. Consider adding a comment.

jcouv commented:
>         public void DefiniteAssignment(bool useBoolReturns, bool trailingOutParameter)

I'm okay to do later. I think the nullability scenarios cover the parts of `AbstractFlowPass` that I was worried about.

---
In reply to: [875184267](https://github.com/dotnet/roslyn/pull/54585#issuecomment-875184267) [](http://example.com/codeflow?ancestors=875184267,874428443)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10243 in 1db9ee9. [](commit_id = 1db9ee9e624d3a853cfad09991c82c2b175e2b68, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -71,64 +70,55 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
                 // Start the sequence with appendProceedLocal, if appropriate
                 BoundExpression? currentExpression = appendShouldProceedLocal;
 
+                var boolType = _compilation.GetSpecialType(SpecialType.System_Boolean);
```
nit (ie. not blocking): we'll want a test where we make the type missing

333fred replied:
Of note, we required this type during initial binding.

jcouv commented:
>                     // (6,5): error CS0165: Use of unassigned local variable 'i'

This seems unexpected when `trailingOutParameter` is false.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10267 in 73be92a. [](commit_id = 73be92aaab4c965102a6e00e2248f5cd34b756f9, deletion_comment = False)

jcouv commented:
>         }

You've covered the definite assignments that I could think of :-)
Just to confirm, usage of argument attribute would not modify affects those scenarios in interesting ways, is that correct?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10310 in 73be92a. [](commit_id = 73be92aaab4c965102a6e00e2248f5cd34b756f9, deletion_comment = False)

jcouv commented:
> ";

nit: all the ref escape tests have this odd spacing (extra blank line)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:11013 in 73be92a. [](commit_id = 73be92aaab4c965102a6e00e2248f5cd34b756f9, deletion_comment = False)

jcouv reviewed (commented):
Done with review pass (iteration 9)

333fred commented:
>                     // (6,5): error CS0165: Use of unassigned local variable 'i'

That would be the failing test.

---
In reply to: [875882504](https://github.com/dotnet/roslyn/pull/54585#issuecomment-875882504) [](http://example.com/codeflow?ancestors=875882504,875882504)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10267 in 73be92a. [](commit_id = 73be92aaab4c965102a6e00e2248f5cd34b756f9, deletion_comment = False)

333fred commented:
>         }

That is correct.

---
In reply to: [875883536](https://github.com/dotnet/roslyn/pull/54585#issuecomment-875883536) [](http://example.com/codeflow?ancestors=875883536,875883536)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10310 in 73be92a. [](commit_id = 73be92aaab4c965102a6e00e2248f5cd34b756f9, deletion_comment = False)

jcouv commented:
>         }

nit: Not blocking, but I'd consider adding a `CustomHandler c = $""{M(out var o)}{o}"";` test with bool returns (should give definite assignment error on second hole/expression).

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10293 in ae6c341. [](commit_id = ae6c34172c080c3dbaa3539b6e60f9a6a5bd7b31, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -1020,7 +1020,11 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
+            var receiverRefKind = receiver?.GetRefKind();
+            uint receiverValEscapeScope = method.RequiresInstanceReceiver
+                ? receiverRefKind?.IsWritableReference() == true ? GetRefEscape(receiver, LocalScopeDepth) : GetValEscape(receiver, LocalScopeDepth)
+                : Binder.ExternalScope;
```
Was this change observable in any test?

333fred replied:
Not that I currently have. I'm honestly not sure how to hit it.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:
```diff
@@ -1261,6 +1261,15 @@ bool isAllowedInitOnlySet(BoundExpression receiver)
+        {
+            Debug.Assert(interpolatedString.InterpolationData != null);
+            var data = interpolatedString.InterpolationData.GetValueOrDefault();
+            return GetValEscape(data.Construction, scopeOfTheContainingExpression);
```
Just to confirm we don't need to deal with `GetRefEscape` here in some cases?
The escapes rules are complicated enough that I'm having trouble reloading this on the fly...

333fred replied:
I don't believe so, but this is not my strongest area.

chsienki replied:
Agree here that I *think* its ok, but I'm not 100% convinced. Can we add a work item to try and make a test that convinces us?

333fred commented:
>         }

This is a success case, not a failure case.

---
In reply to: [875936424](https://github.com/dotnet/roslyn/pull/54585#issuecomment-875936424) [](http://example.com/codeflow?ancestors=875936424,875936424)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10293 in ae6c341. [](commit_id = ae6c34172c080c3dbaa3539b6e60f9a6a5bd7b31, deletion_comment = False)

jcouv reviewed (approved):
LGTM Thanks (iteration 11). Remaining test questions/suggestions can be tracked for later.

jcouv commented:
>         }

Doh. Thanks :-)

---
In reply to: [875939910](https://github.com/dotnet/roslyn/pull/54585#issuecomment-875939910) [](http://example.com/codeflow?ancestors=875939910,875936424)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10293 in 73be92a. [](commit_id = 73be92aaab4c965102a6e00e2248f5cd34b756f9, deletion_comment = False)

chsienki commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1973,29 +1988,140 @@ internal IOperation CreateBoundTupleOperation(BoundTupleExpression boundTupleExp
         {
-            var builder = ArrayBuilder<IInterpolatedStringContentOperation>.GetInstance(parts.Length);
-            foreach (var part in parts)
+            return data is { PositionInfo: var positionInfo } ? createHandlerInterpolatedStringContent(positionInfo) : createNonHandlerInterpolatedStringContent();
```
Nit: personally I find these `a` or `b` methods with a pair of local functions harder to follow logically. Also the extra method call could be inefficient. (Purely personal, but just something to consider)

333fred replied:
> Also the extra method call could be inefficient.

It's small enough that the JIT will likely just inline the containing call :).

chsienki commented on src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:
```diff
@@ -1105,7 +1105,10 @@ public override void VisitInterpolatedString(IInterpolatedStringOperation operat
         {
             Assert.Equal(OperationKind.InterpolatedStringText, operation.Kind);
-            Assert.Equal(OperationKind.Literal, operation.Text.Kind);
+            if (operation.Text.Kind != OperationKind.Literal)
```
Stupid question: why do we know its not a literal? because it could be dynamic?

333fred replied:
We don't know it's not a literal. If it is a literal, everything is fine. If it's not a literal, it must be a conversion wrapping a literal.

jcouv commented:
nit: It looks like this PR was merged instead of squashed :-/
